### PR TITLE
API: shuffle dask array

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/dask/dask/issues/3409

Some brief timings on

```python
x = da.random.random((100_000, 10), chunks=10_000)
index = np.arange(len(x))
np.random.shuffle(index)
```

task | old | new
---- | --- | ---
build graph | 419 ms | 28.7 ms
compute | 14.5 s | 48.9 ms

### CODE REVIEW

**Comment by TomAugspurger** [comment]
IIUC, In #3409, @mrocklin mentioned adjusting `slicing_plan` to detect when we should use this slicing method. I've not attempted that. So a "naive" shuffle of a dask array with

```python
index = np.arange(len(arr))
np.random.shuffle(index)
arr[index]
```

is still going to be very slow. But the use cases I have in mind (#3409, https://github.com/lmcinnes/umap/issues/62, approximate nearest neighbors) can opt into the faster slicing, when we know we have the right kind of `index` array.

**Comment by mrocklin** [comment]
There might be some valuable code in https://github.com/dask/dask/pull/3808
for this operation.

**Comment by TomAugspurger** [comment]
Thanks, I had forgotten about that. I'll take a look.

File: dask/array/slicing.py [hunk 1]
```diff
@@ -1033,3 +1033,80 @@ def slice_with_bool_dask_array(x, index):
...
+    from .core import slices_from_chunks
+
+    slices = slices_from_chunks(chunks)
+
+    if len(slices[0]) > 1:
+        slices = [slice_[0] for slice_ in slices]
+
+    offsets = np.roll(np.cumsum(chunks[0]), 1)
+    offsets[0] = 0
+    offsets
```
Comment (jakirkham) [inline dask/array/slicing.py]: Is this needed?

**Comment by jcrist** [comment]
@TomAugspurger, what's left to be done here?

**Comment by TomAugspurger** [comment]
Sorry, forgot about this. I think that this is useful, so I've fixed the merge conflicts. I haven't re-reviewed the implementation though.

**Comment by jakirkham** [comment]
@shoyer, any thoughts on this implementation of shuffle for Dask Arrays?

**Comment by martindurant** [comment]
Ping: this seems to have been left to go stale

**Comment by TomAugspurger** [comment]
The difference in inplace vs. a new Array is the main thing concerning me. Do we have other places in dask.array that differ from NumPy like this?

**Comment by martindurant** [comment]
So long as the doc is clear, it should be OK -  we are different from numpy in a number of ways in a number of places; I expect that includes in-place behaviour somewhere, although I don't know for sure.

File: dask/array/random.py [hunk 2]
```diff
@@ -348,6 +348,48 @@ def random_sample(self, size=None, chunks="auto"):
     def rayleigh(self, scale=1.0, size=None, chunks="auto"):
         return self._wrap('rayleigh', scale, size=size, chunks=chunks)
 
+    def shuffle(self, x):
```
Comment (shoyer) [inline dask/array/random.py]: Why not call this `permutation` instead, which is consistent with the naming in numpy.random?

File: dask/array/tests/test_random.py [hunk 3]
```diff
@@ -293,6 +293,17 @@ def test_names():
     assert len(key_split(name)) < 10
 
 
+def test_shuffle():
+    x = da.arange(12, chunks=3)
+    da.random.shuffle(x)
```
Comment (shoyer) [inline dask/array/tests/test_random.py]: Does this line do anything?

File: dask/array/random.py (hunk 2 above)
Comment (TomAugspurger) [reply dask/array/random.py]: Oh, I'd forgotten that permutation can take an array! That's much nicer.

I suspect we can also support the integer variant. I'll pick this up later this week or next.

**Comment by jcrist** [comment]
> The difference in inplace vs. a new Array is the main thing concerning me.

There are other places in the api where we mutate an existing array/dataframe object inplace (this is fine as long as we don't mutate the graph). I'd prefer we match numpy's mutating api here if possible, as I suspect differing will lead to user issues in the future.

**Comment by TomAugspurger** [comment]
Merging later today if there aren't any objections.

**Comment by jakirkham** [comment]
Thanks for working on this Tom! 😄

**Comment by stsievert** [comment]
Thanks for working on this @TomAugspurger!

I've rerun the timing comparison to see how the new implementation works and show how to use it.

| Implementation | Graph build | Computation |
|:---:|:--:|:--:|
| `shuffle_blocks` (this PR) |  77.8ms | 67.1 ms |
| Naive indexing | 721ms | 13.007s |

Here's the code I used to generate it:
<details>

``` python
import dask.array as da
import numpy as np
import dask
from time import time
from dask.array.slicing import shuffle_slice

if __name__ == "__main__":
    x = da.random.random((100_000, 10), chunks=10_000)
    index = np.arange(len(x))
    np.random.shuffle(index)

    start = time()
    y2 = shuffle_slice(x, index)  # 0.07785s
    print(time() - start)
    start = time()
    z2 = y2.compute()  # 0.06716
    print(time() - start)

    start = time()
    y1 = x[index]  # 0.721
    print(time() - start)
    start = time()
    z1 = y1.compute()  # 13.0067
    print(time() - start)

```

</details>

The core of this code is

``` python
x = da.random.random((100_000, 10), chunks=10_000)
index = np.arange(len(x))
np.random.shuffle(index)

y1 = shuffle_blocks(x, index)  # shuffle_blocks
y2 = x[index]  # naive indexing
```

**Pull request merged**